File: Design-Patterns-Using-Ruby/ArchitecturalPatterns/MVC.rb
# Model
class Model
    attr_accessor :data

    def initialize
        @data = nil
        @observers = []
    end

    def data=(value)
        puts 'Model : Set data.'
        @data = value
        notify_observers
    end

    def add_observer(observer)
        @observers << observer
    end

    def remove_observer(observer)
        @observers.delete(observer)
    end

    def notify_observers
        puts 'Model : Notify observers.'
        @observers.each(&:update)
    end
end

# View
class View
    def initialize(model, controller)
        @model = model
        @controller = controller
        @model.add_observer(self)
    end

    def update
        puts 'View : Update.'
        data = @model.data
        puts "Data: #{data}"
    end

    def get_user_input
        print 'View : Enter user input: '
        # user_input = gets.chomp
        user_input = 'Hello, World!'
        puts user_input
        @controller.handle_user_input(user_input)
    end
end

# Controller
class Controller
    def initialize(model)
        @model = model
    end

    def handle_user_input(user_input)
        puts 'Controller : Handle user input.'
        @model.data = user_input
    end

    def set_view(view)
        @view = view
    end
end

# Main
model = Model.new
controller = Controller.new(model)
view = View.new(model, controller)
controller.set_view(view)
view.get_user_input

=begin 
View : Enter user input: Hello, World!
Controller : Handle user input.
Model : Set data.
Model : Notify observers.
View : Update.
Data: Hello, World! 
=end

File: Design-Patterns-Using-CSharp/ArchitecturalPatterns/MVC.cs
using System;
using System.Collections.Generic;

// Model
class Model
{
    private string data;
    private List<View> observers;

    public Model()
    {
        this.observers = new List<View>();
    }

    public void SetData(string data)
    {
        Console.WriteLine("Model: Set data.");
        this.data = data;
        NotifyObservers();
    }

    public string GetData()
    {
        Console.WriteLine("Model: Get data.");
        return this.data;
    }

    public void AddObserver(View observer)
    {
        this.observers.Add(observer);
    }

    public void RemoveObserver(View observer)
    {
        this.observers.Remove(observer);
    }

    public void NotifyObservers()
    {
        Console.WriteLine("Model: Notify observers.");
        foreach (View observer in observers)
        {
            observer.Update();
        }
    }
}

// View
class View 
{
    private Controller controller;
    private Model model;

    public View(Model model, Controller controller)
    {
        this.model = model;
        this.controller = controller;
        this.model.AddObserver(this);
    }

    public void Update()
    {
        Console.WriteLine("View: Update.");
        string data = model.GetData();
        Console.WriteLine("Data: " + data);
    }

    public void GetUserInput()
    {
        Console.Write("View: Enter user input: ");
        string userInput = Console.ReadLine();
        controller.HandleUserInput(userInput);
    }
}

// Controller
class Controller
{
    private Model model;
    private View view;

    public Controller(Model m)
    {
        this.model = m;
    }

    public void HandleUserInput(string userInput)
    {
        Console.WriteLine("Controller: Handle user input.");
        model.SetData(userInput);
        // Can inform view about action.
    }

    public void SetView(View v)
    {
        this.view = v;
    }
}

// Main class
class MVC
{
    public static void Main(string[] args)
    {
        Model model = new Model();
        Controller controller = new Controller(model);  // The Controller sets itself as the observer in this case
        View view = new View(model, controller);
        controller.SetView(view);
        view.GetUserInput();
    }
}

/*
View: Enter user input: hello, world!
Controller: Handle user input.
Model: Set data.
Model: Notify observers.
View: Update.
Model: Get data.
Data: hello, world!
*/
